### PR TITLE
filters: add proper support for hardware format conversion

### PR DESCRIPTION
**Pull request by philipl**
This series of changes was originally part of my Vulkan hwdec interop PR. But as the actual Vulkan support is blocked on ffmpeg changes, I'm breaking out the filter parts to be merged separately.

The main motivator for this series is to allow for the use of ffmpeg's Vulkan based filters in conjunction with hardware decoding and hardware rendering. In these scenarios, we must transfer from one hardware format to another and possibly back again. That can only work if hardware contexts can be properly initialised and assigned to the right filters.

Note that you can't use these new capabilities for Vulkan filters without the Vulkan interop because the hw context will not be initialised correctly by default, but I have tested it with that pending set of changes.

Verified combinations:
- [x] cuda -> vulkan
- [x] cuda -> vulkan -> cuda
- [x] vaapi -> vulkan
- [x] vaapi -> vulkan -> vaapi

### CODE REVIEW

File: filters/f_hwtransfer.c [hunk 1]
```diff
@@ -58,6 +62,33 @@ static const struct ffmpeg_and_other_bugs shitlist[] = {
     {0}
 };
 
+struct hwmap_pairs_t {
```
Comment (sfan5) [inline filters/f_hwtransfer.c:65]: `_t` types are reserved by the standard library

File: filters/f_lavfi.c [hunk 2]
```diff
@@ -548,11 +552,22 @@ static void init_graph(struct lavfi *c)
...
         struct mp_stream_info *info = mp_filter_find_stream_info(c->f);
         if (info && info->hwdec_devs) {
-            struct mp_hwdec_ctx *hwdec = hwdec_devices_get_first(info->hwdec_devs);
-            for (int n = 0; n < c->graph->nb_filters; n++) {
-                AVFilterContext *filter = c->graph->filters[n];
-                if (hwdec && hwdec->av_device_ref)
-                    filter->hw_device_ctx = av_buffer_ref(hwdec->av_device_ref);
+            struct mp_hwdec_ctx *hwdec_ctx = NULL;
+            if (c->hwdec_interop) {
+                MP_VERBOSE(c, "Configuring hwdec_interop=%s for filters\n",
```
[the remark this answers is not in the document]
Comment (philipl) [reply filters/f_lavfi.c:557]: As in use `MP_INFO`?

[the remark this answers is not in the document]
Comment (philipl) [reply filters/f_lavfi.c:557]: Or you mean, if we end up just loading the first one, print that too?

[the remark this answers is not in the document]
Comment (sfan5) [reply filters/f_lavfi.c:557]: latter

**Pull request merged**